dataset/src/main/scala/frameless/TypedColumn.scala

```diff
@@ -27,31 +59,32 @@ sealed trait UntypedExpression[T] {
   * @tparam T type of dataset
   * @tparam U type of column
   */
-sealed class TypedColumn[T, U](
-  val expr: Expression)(
-  implicit
-  val uencoder: TypedEncoder[U]
+abstract class AbstractTypedColumn[T, U](val expr: Expression)(
+  implicit val uencoder: TypedEncoder[U]
 ) extends UntypedExpression[T] { self =>
 
-  /** From an untyped Column to a [[TypedColumn]]
-    *
-    * @param column a spark.sql Column
-    * @param uencoder encoder of the resulting type U
-    */
-  def this(column: Column)(implicit uencoder: TypedEncoder[U]) {
-    this(FramelessInternals.expr(column))
-  }
+  type TC[A] <: AbstractTypedColumn[T, A]
```

Comment:
Both of these are things I considered. No particular reason on taking this approach other that it simplifies the type signatures throughout. It does exposes less, so given that you have already a use case that needs T, I can make these changes.

Reply:
@OlivierBlanvillain I didn't find a nice way to do the same I do with using a type member. Any suggestions? It gets into this recursive type definition and finding myself to use a type-lamda. This is what I have:

```scala
AbstractTypedColumn[T, U, TC[T, _] <: AbstractTypedColumn[T, U, _]]
```

But I would want to write something like:

```scala
AbstractTypedColumn[T, U, TC[T,?] <: AbstractTypedColumn[T, U, ?]]
```

Idk, unless you have any other ideas, I think type member looks much better for doing this than a type parameter.

Reply:
I guess you could just write `TC[_, _]`, no need to constrain it further given than it's for internal use only.

Reply:
I take that back, you obviously need the the `<: AbstractTypedColumn` to be able to do call methods on `TC`... Here is the type parameter version:

```scala
def AbstractTypedColumn
  [T, U, ThisType[x, y] <: AbstractTypedColumn[x, y, ThisType]] ...
```

I see no benefits of using that instead of the type member version, so let's stick to what's already here.